[SNAP-2032] move bucket size tracking inside RegionEntry._setValue

sumwale opened the pull request:
## Changes proposed in this pull request

- to make size tracking much more reliable, moved the size tracking to be inside
  RegionEntry._setValue
- extra argument of RegionEntryContext for tracking so signatures of callers changed likewise
- BucketRegion.updateMemoryStats method taking old+new value that is invoked from above;
  the previous entry point of updateBucket2Size is now commented out
- separate call to updateMemoryStats in AbstractRegionMap.copyRecoveredEntries since the
  RegionEntryContext changes here from place holder region to actual region
- always faultin old value for internal Delta like for public Delta; this avoids multiple reads
  of old value without faultin (by NonLocalRegionEntry creation and then by updateEntry)
- fixed accounting error in UMM where overhead was added in delete not in update (as
  required) but freePoolMemory was always reducing entry overhead; now storing the original
  calculated size as well as op=DELETE flag in NonLocalRegionEntry which are used in free
- allow explicit call to old map cleaner task, and make it thread-safe
- reduced old map cleaner time to 20s
- added buildPrefix to build.gradle so that store targets can be run independently

## Patch testing

precheckin and manual TPCH testing with large bulk deletes

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/856

Commit:
[SNAP-2032] move bucket size tracking inside RegionEntry._setValue

- to make size tracking much more reliable, moved the size tracking to be inside RegionEntry._setValue
- extra argument of RegionEntryContext for tracking so signatures of callers changed likewise
- new BucketRegion.updateMemoryStats method taking old and new value that is invoked from above;
  the previous entry point of updateBucket2Size is now commented out
- separate call to updateMemoryStats in AbstractRegionMap.copyRecoveredEntries since the
  RegionEntryContext changes here from place holder region to actual region
- fixed accounting error in UMM where overhead was added in delete not in update (as required)
  but freePoolMemory was reducing entry overhead in all cases; now storing the original calculated
  size as well as op=DELETE flag in NonLocalRegionEntry which are used in free
- allow explicit call to old map cleaner task, and make it thread-safe
- reduced old map cleaner time to 20s
- added buildPrefix to build.gradle so that store targets can be run independently

Commit:
missed in last

Commit:
comment update

rishitesh commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -605,7 +605,7 @@ public void addOldEntry(RegionEntry oldRe, RegionEntry newEntry, LocalRegion reg
       region.calculateEntryOverhead(oldRe);
       LocalRegion.regionPath.set(region.getFullPath());
-      region.acquirePoolMemory(0, oldSize, forDelete, null, true);
+      region.acquirePoolMemory(0, oldRe.getValueSize(), oldRe.isForDelete(), null, true);
```
My assumption was we will always create a new entry while copying to old entry map. Hence earlier it was true for both while acquiring and releasing. Even for update we must be creating a new entry while copying ?

sumwale replied:
@rishitesh If that's what you want to account then need to add the overhead of NonLocalRegionEntry. However, the update call to acquirePool was skipping overhead completely (the "forDelete" flag) while free was adding. Perhaps can add NonLocalRegionEntry overhead fo oldRe.valueSize before acquire call then it will get accounted in both places now.

rishitesh commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -605,7 +605,7 @@ public void addOldEntry(RegionEntry oldRe, RegionEntry newEntry, LocalRegion reg
       region.calculateEntryOverhead(oldRe);
       LocalRegion.regionPath.set(region.getFullPath());
-      region.acquirePoolMemory(0, oldSize, forDelete, null, true);
+      region.acquirePoolMemory(0, oldRe.getValueSize(), oldRe.isForDelete(), null, true);
```
My assumption was we will always create a new entry while copying to old entry map. Hence earlier it was true for both while acquiring and releasing. Even for update we must be creating a new entry while copying ?

rishitesh commented:
looks fine. one comment.

suranjan commented:
Looks good.

sumwale commented:
@rishitesh Did you comment somewhere? Can't see it here.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractRegionMap.java:
```diff
@@ -847,8 +849,12 @@ public void copyRecoveredEntries(RegionMap rm, boolean entriesIncompatible) {
-              _getOwner().calculateRegionEntryValueSize(oldRe));
+          // only for incrementing count while size is updated below
+          owner.updateSizeOnCreate(key, 0);
+          if (!oldRe.isOffHeap()) {
```
Just a doubt. Is there a possibility that since accounting has moved to _setValue then in case of changeOwner where just map is borrowed from PlaceHolderDiskRegion there can be a scope of double accounting.

sumwale replied:
@kneeraj For PlaceHolderDiskRegion the updateMemoryStats is empty so there will be no accounting done then. I have cleaned up this a bit to move the accounting inside AbstractRegionEntry.setOwner itself -- please see https://github.com/SnappyDataInc/snappy-store/pull/316/commits/dc8c5d0bd952f89cee7ae8a0a53f4791585114be

kneeraj reviewed (commented):
Looks good. One clarification seeked.

rishitesh reviewed (commented):
My assumption was we will always create a new entry while copying to old entry map. Hence earlier it was true for both while acquiring and releasing. Even for update we must be creating a new entry while copying ?

rishitesh commented:
@sumwale something happened and my comments now got posted three times.

Commit:
Moved ownership change accounting inside AbstractRegionEntry itself

Removes unnecessary assumption and avoids missing calls from other routes to setOwner.
